src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +69,101 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = []  # type: List[InstallationResult]
+
+    # store install arguments
+    install_args = _InstallArgs(
+        install_options=install_options,
+        global_options=global_options,
+        root=root,
+        home=home,
+        prefix=prefix,
+        warn_script_location=warn_script_location,
+        use_user_site=use_user_site,
+        pycompile=pycompile,
+    )
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
+        # get a list of packages we can install in parallel
+        parallel_reqs_index = [
+            i for i, req in enumerate(requirements)
+            if not req.should_reinstall and req.is_wheel]
+
+        # install packages parallel
+        if parallel_reqs_index:
+            parallel_reqs = map_multiprocess_ordered(
+                partial(_single_install, install_args, suppress_exception=True),
+                [requirements[i] for i in parallel_reqs_index])
```

Comment:
I’d probably make this a `Dict[index, result]` instead. It’d be easier to get the results out in the later loop (`parallel_reqs[i]` and catch `KeyError`).

Reply:
You mean like:
```python
parallel_reqs_dict = dict(zip(parallel_reqs_index, parallel_reqs))
...
try:
    installed_req = parallel_reqs_dict[i]
except KeyError:
    installed_req = _single_install(install_args, req, suppress_exception=False)
```

Reply:
The `dict(zip())` line looks wrong to me. I’m thinking something like

```python
def map_multiprocess(func, keys, values, chunksize=1):
    with closing(ProcessPool()) as pool:
        mapped = pool.map(func, values, chunksize)
    return dict(zip(keys, mapped))

should_parallel_reqs = (
    (i, req)
    for i, req in enumerate(requirements)
    if not req.should_reinstall and req.is_wheel
)
should_parallel_indexes, should_parallel_values = zip(*should_parallel_reqs)

parallel_reqs_dict = map_multiprocess(partial(...), should_parallel_indexes, should_parallel_values)

for i, req in enumerate(requirements):
    try:
        installed_req = parallel_reqs_dict[i]
    except KeyError:
        # Fallback...
```

Reply:
Sure, I'll change a bit so we use the internal `_map_multiprocess_ordered`, something like
```python
parallel_reqs_dict = dict(
    zip(should_parallel_indexes,
        map_multiprocess_ordered(
            partial(_single_install,
                    install_args,
                    suppress_exception=True),
            should_parallel_values)))

```
looks okay to you?

Reply:
The logic looks good to me, but I’d probably split it into at least two statements (one map and one zip-dict). This is pretty difficult to unpack.

Reply:
done